src/impl/handover.cpp

```diff
@@ -0,0 +1,102 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#include "handover.hpp"
+
+using namespace realm;
+using namespace realm::_impl;
+
+AnyHandover::AnyHandover(AnyHandover&& handover)
+{
+    switch (handover.m_type) {
+        case AnyThreadConfined::Type::Object:
+            new (&m_object.row_handover) RowHandover(std::move(handover.m_object.row_handover));
+            new (&m_object.object_schema_name) std::string(std::move(handover.m_object.object_schema_name));
+            break;
+
+        case AnyThreadConfined::Type::List:
+            new (&m_list.link_view_handover) LinkViewHandover(std::move(handover.m_list.link_view_handover));
+            break;
+
+        case AnyThreadConfined::Type::Results:
+            new (&m_results.query_handover) QueryHandover(std::move(handover.m_results.query_handover));
+            new (&m_results.sort_order) SortOrder(std::move(handover.m_results.sort_order));
+            break;
+    }
+    new (&m_type) AnyThreadConfined::Type(handover.m_type);
+}
+
+AnyHandover& AnyHandover::operator=(AnyHandover&& handover)
+{
+    switch (handover.m_type) {
```

Comment:
This implementation doesn't appear to handle destroying the existing state before it initializes the new state.